src/crdt_rga.erl

```diff
@@ -0,0 +1,298 @@
+generate_downstream({addRight, Elem, Position}, _Actor, Rga) ->
+    case (Position < 0) or (Position > length(Rga)) of
+        true -> {error, {invalid_position, Position}};
+        false -> case length(Rga) of
+                     0 -> {ok, {addRight, {ok, 0, 0}, {ok, Elem, unique()}}};
+                     _ -> {ok, {addRight, lists:nth(Position, Rga), {ok, Elem, unique()}}}
+                 end
+    end;
+generate_downstream({remove, Position}, _Actor, Rga) ->
+    {ok, {remove, lists:nth(Position, Rga)}}.
+
+%% @doc given an RGA, returns the same RGA, as it represents its own value.
+-spec value(rga()) -> rga().
+value(Rga) ->
+    Rga.
+
+%% @doc This method takes in an rga operation and an rga to perform the operation.
+%% It returns either the added Vertex and the new rga in case of an addRight, and the new rga in the case of a remove.
+-spec update(rga_downstream_op(), rga()) -> rga_result().
+update({addRight, RightVertex, NewVertex}, Rga) ->
+    recursive_insert(RightVertex, NewVertex, Rga, []);
+update({remove, Vertex}, Rga) ->
+    remove_vertex(Vertex, Rga).
+
+%% Private
+%% @doc recursively looks for the Vertex where the new element should be put to the right of.
+-spec recursive_insert(vertex(), vertex(), rga(), list()) -> rga_result().
+recursive_insert(_, NewVertex, [], []) ->
+    {ok, [NewVertex]};
+recursive_insert({ok, 0, 0}, NewVertex, L, []) ->
+    add_element(NewVertex, L, []);
+recursive_insert(RightVertex, NewVertex, [RightVertex | T], L) ->
+    add_element(NewVertex, T, [RightVertex | L]);
+recursive_insert(RightVertex, NewVertex, [H | T], L) ->
+    recursive_insert(RightVertex, NewVertex, T, [H | L]).
+
+%% Private
+%% @doc the place for the insertion has been found, so now the UIDs are compared to see where to insert.
+-spec add_element(vertex(), rga(), list()) -> rga_result().
+add_element({Status, Value, UID}, [{Status1, Value1, UID1} | T], L) ->
+    case UID >= UID1 of
+        true ->
+            {ok, lists:reverse(L) ++ [{Status, Value, UID}] ++ [{Status1, Value1, UID1} | T]};
+        _ ->
+            add_element({Status, Value, UID}, T, [{Status1, Value1, UID1} | L])
+    end;
+add_element(Insert, [], L) ->
+    {ok, lists:reverse([Insert | L])}.
+
+%% Private
+%% @doc looks for the Vertex to be removed. Once it's found, it's marked as "deleted".
+%% The Vertex is not removed from the list, to allow adding elements to its right.
+-spec remove_vertex(vertex(), rga()) -> rga_result().
+remove_vertex({ok, Value, UID}, Rga) ->
+    {ok, lists:keyreplace(UID, 3, Rga, {deleted, Value, UID})}.
+
+%% @doc given an rga, this mehtod looks for all tombstones and removes them, returning the tombstone free rga.
+-spec purge_tombstones(rga()) -> rga_result().
+purge_tombstones(Rga) ->
+    L = lists:filter(fun({Status, _, _}) ->
+        Status == ok
+    end, Rga),
+    {ok, L}.
+
+%% @doc generate a unique identifier (best-effort).
+unique() ->
+    crypto:strong_rand_bytes(20).
```

Comment:
Ah, there is this also:
http://www.erlang.org/doc/man/erlang.html#unique_integer-1
maybe
{node(),erlang:unique_integer([])}

is better than make_ref() because I'm not sure how the reference is stored exactly.